src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,344 @@
+      try {
+        wait(delay - elapsed);
+      } catch (InterruptedException ex) {
+      }
+      elapsed = System.currentTimeMillis() - start;
+    }
+    return running;
+  }
+
+  private List<String> readChunk(Optional<String> from, int chunkSize) {
+    try (Timer.Context timer = readChunkTimer.time()) {
+      if (from.isPresent()) {
+        return accountsManager.getAllNumbers(from.get(), chunkSize);
+      } else {
+        return accountsManager.getAllNumbers(chunkSize);
+      }
+    }
+  }
+
+  private Optional<Response> sendChunk(Optional<String> fromNumber, List<String> numbers) {
+    String reconcilePath;
+    if (fromNumber.isPresent()) {
+      reconcilePath = String.format("/v1/directory/reconcile/%s", fromNumber);
+    } else {
+      reconcilePath = "/v1/directory/reconcile";
+    }
+
+    try (Timer.Context timer = sendChunkTimer.time()) {
+      Response response = client.target(serverApiUrl)
+                                .path(reconcilePath)
+                                .request(MediaType.APPLICATION_JSON)
+                                .put(Entity.json(new DirectoryReconciliationRequest(numbers)));
+      if (response.getStatus() != 200) {
+        sendChunkErrorMeter.mark();
+        logger.warn("http error " + response.getStatus());
+      }
+      return Optional.of(response);
+    } catch (ProcessingException ex) {
+      sendChunkErrorMeter.mark();
+      logger.warn("request error: ", ex);
+      return Optional.absent();
+    }
+  }
+
+  private void processChunk(Jedis jedis) {
+    Optional<String> fromNumber     = Optional.fromNullable(jedis.get(LAST_NUMBER_KEY));
+    Optional<String> cachedCountStr = Optional.fromNullable(jedis.get(CACHED_COUNT_KEY));
+    Optional<Long>   cachedCount    = Optional.absent();
+    if (cachedCountStr.isPresent()) {
+      cachedCount = tryParseLong(cachedCountStr.get());
+    }
+
+    if (!cachedCount.isPresent()) {
+      cachedCount = Optional.of(accountsManager.getCount());
+      jedis.psetex(CACHED_COUNT_KEY, CACHED_COUNT_TTL_MS, Long.toString(cachedCount.get()));
+    }
+
+    int          chunkSize = (int) (cachedCount.get() * PERIOD / CHUNK_INTERVAL);
+    List<String> numbers   = readChunk(fromNumber, chunkSize);
+
+    Optional<String> toNumber = Optional.absent();
+    if (!numbers.isEmpty()) {
+      toNumber = Optional.of(numbers.get(numbers.size() - 1));
+    }
+
+    Optional<Response> sendChunkResp = sendChunk(fromNumber, numbers);
+    if (sendChunkResp.isPresent()) {
+      if (sendChunkResp.get().getStatus() == 200) {
+        if (toNumber.isPresent()) {
+          jedis.set(LAST_NUMBER_KEY, toNumber.get());
+        } else {
+          jedis.del(ACCELERATE_KEY);
+          jedis.del(LAST_NUMBER_KEY);
+        }
+      } else {
+        jedis.del(ACCELERATE_KEY);
+        if (sendChunkResp.get().getStatus() == 404) {
+          jedis.del(LAST_NUMBER_KEY);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void run() {
```

Comment:
I am currently unable to follow the logic of this method. Is it possible to try refactoring it so that it reads more like a book of readable instructions? The pattern that I like to follow is that public methods call private methods which perform operations on their inputs, without modifying instance state, and return an output that is coordinated from the public method into the next private method. It's also possible that a test might help me understand how this is supposed to work.

Reply:
yeah, sorry, i pushed this code in a rush before flying in case my laptop got lost.. i'll let you know when it's ready for review again.

Reply:
ok PTAL, rewrote the code to be more readable / testable in 704f9c84136e

Reply:
tests added.. ready for review in 7e6ffa5